src/main/scala/org/broadinstitute/dsde/workbench/sam/service/UserService.scala

```diff
@@ -84,15 +86,50 @@ class UserService(val directoryDAO: DirectoryDAO, val googleDirectoryDAO: Google
     } yield deleteResult
   }
 
+  def createUserPetServiceAccount(user: WorkbenchUser): Future[WorkbenchUserPetServiceAccountEmail] = {
+    def create() = {
+      val (petSa, petSaDisplayName) = toPetSAFromUser(user)
+      for {
+        petServiceAccount <- googleIamDAO.createServiceAccount(petServiceAccountConfig.googleProject, petSa, petSaDisplayName)
+        _ <- googleDirectoryDAO.addMemberToGroup(WorkbenchGroupEmail(toProxyFromUser(user.id.value)), petServiceAccount.email)
+        _ <- Future.traverse(petServiceAccountConfig.serviceAccountActors) { email =>
+          googleIamDAO.addServiceAccountActorRoleForUser(petServiceAccountConfig.googleProject, petServiceAccount.email, email)
+        }
+        _ <- directoryDAO.addPetServiceAccountToUser(user.id, petServiceAccount.email)
+      } yield petServiceAccount.email
+    }
+
+    directoryDAO.getPetServiceAccountForUser(user.id).flatMap {
+      case Some(email) => Future.successful(email)
+      case None => create()
+    }
+  }
+
   private def createAllUsersGroup: Future[Unit] = {
     for {
       _ <- directoryDAO.createGroup(WorkbenchGroup(allUsersGroupName, Set.empty, WorkbenchGroupEmail(toGoogleGroupName(allUsersGroupName.value)))) recover { case e: WorkbenchExceptionWithErrorReport if e.errorReport.statusCode == Option(StatusCodes.Conflict) => () }
       _ <- googleDirectoryDAO.createGroup(WorkbenchGroupName(allUsersGroupName.value), WorkbenchGroupEmail(toGoogleGroupName(allUsersGroupName.value))) recover { case e: GoogleJsonResponseException if e.getDetails.getCode == StatusCodes.Conflict.intValue => () }
     } yield ()
   }
 
-  private def toProxyFromUser(subjectId: String): String = s"PROXY_$subjectId@$googleDomain"
-  private def toGoogleGroupName(groupName: String): String = s"GROUP_$groupName@$googleDomain"
+  private[service] def toProxyFromUser(subjectId: String): String = s"PROXY_$subjectId@$googleDomain"
+  private[service] def toGoogleGroupName(groupName: String): String = s"GROUP_$groupName@$googleDomain"
+
+  private[service] def toPetSAFromUser(user: WorkbenchUser): (WorkbenchUserPetServiceAccountId, WorkbenchUserPetServiceAccountDisplayName) = {
+    /*
+     * Service account IDs must be:
+     * 1. between 6 and 30 characters
+     * 2. lower case alphanumeric separated by hyphens
+     * 3. must start with a lower case letter
+     *
+     * UUID.randomUUID() is too many characters, so generate a random 8 byte UUID prefixed with "pet-".
+     */
+    val randLong = ByteBuffer.wrap(UUID.randomUUID().toString.getBytes()).getLong
+    val serviceAccountId = s"pet-${java.lang.Long.toString(randLong, Character.MAX_RADIX)}"
```

Comment:
is there a reason not to use the user's subject id? It is 22 chars long. We already use it for the proxy.

Reply:
Ah, that's probably a good idea :)